File: src/rb/ThriftClient/hypertable/thriftclient.rb
# include path hack
$:.push(File.dirname(__FILE__) + '/gen-rb')

require 'thrift'
require 'thrift/protocol/binaryprotocolaccelerated'
require 'HqlService'

module Hypertable
  class ThriftClient < ThriftGen::HqlService::Client
    def initialize(host, port, timeout_ms = 20000, do_open = true)
      @transport = Thrift::FramedTransport.new(Thrift::Socket.new(host, port))
      protocol = Thrift::BinaryProtocolAccelerated.new(@transport)
      super(protocol)
      open() if do_open
    end

    def open()
      @transport.open()
      @do_close = true
    end

    def close()
      @transport.close() if @do_close
    end

    # more convenience methods

    # buffered query
    def hql_query(hql)
      hql_exec(hql, false, false);
    end

    def with_scanner(table, scan_spec)
      scanner = open_scanner(table, scan_spec)
      yield scanner
      close_scanner(scanner)
    end

    def with_mutator(table)
      mutator = open_mutator(table);
      yield mutator
      close_mutator(mutator)
    end

    # scanner iterator
    def each_cell(scanner)
      cells = next_cells(scanner);

      while (cells.size)
        cells.each {|cell| yield cell}
        cells = next_cells(scanner);
      end
    end

  end
end

File: src/py/ThriftClient/hypertable/thriftclient.py
from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

from hyperthrift.gen2 import HqlService

class ThriftClient(HqlService.Client):
  def __init__(self, host, port, timeout_ms = 300000, do_open = 1):
    socket = TSocket.TSocket(host, port)
    socket.setTimeout(timeout_ms)
    self.transport = TTransport.TFramedTransport(socket)
    protocol = TBinaryProtocol.TBinaryProtocol(self.transport)
    HqlService.Client.__init__(self, protocol)

    if do_open:
      self.open(timeout_ms)

  def open(self, timeout_ms):
    self.transport.open()
    self.do_close = 1

  def close(self):
    if self.do_close:
      self.transport.close()
